[AMDGPU] Skip register uses in AMDGPUResourceUsageAnalysis

rovka opened the pull request:
Don't count register uses when determining the maximum number of
registers used by a function. Count only the defs. This is really an
underestimate of the true register usage, but in practice that's not
a problem because if a function uses a register, then it has either
defined it earlier, or some other function that executed before has
defined it.

In particular, the register counts are used:
1. When launching an entry function - in which case we're safe because
   the register counts of the entry function will include the register
   counts of all callees.
2. At function boundaries in dynamic VGPR mode. In this case it's safe
   because whenever we set the new VGPR allocation we take into account
   the outgoing_vgpr_count set by the middle-end.

The main advantage of doing this is that the artificial VGPR arguments
used only for preserving the inactive lanes when using the
llvm.amdgcn.init.whole.wave intrinsic are no longer counted. This
enables us to allocate only the registers we need in dynamic VGPR mode.

rovka committed:
[AMDGPU] Ignore inactive VGPRs in .vgpr_count

When using the amdgcn.init.whole.wave intrinsic, we add dummy VGPR
arguments with the purpose of preserving their inactive lanes. The
pattern may look something like this:

```
entry:
  call amdgcn.init.whole.wave
  brand to shader or tail

shader:
  $vInactive = IMPLICIT_DEF ; Tells regalloc it's safe to use the active lanes
  actual code...

tail:
  call amdgcn.cs.chain [...], implicit $vInactive
```

We should not report these VGPRs in the .vgpr_count metadata. This patch
achieves that goal by ignoring IMPLICIT_DEFs and SI_TCRETURNs in
functions that use the amdgcn.init.whole.wave intrinsic. All other VGPRs
are counted as usual.

jayfoad commented:
This could be much more generic. AMDGPUResourceUsageAnalysis could unconditionally ignore IMPLICIT_DEF and all use operands.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUResourceUsageAnalysis.cpp:
```diff
@@ -239,6 +245,16 @@ AMDGPUResourceUsageAnalysis::analyzeResourceUsage(
+        // VGPR usage. We can identify them easily because they're only used in
+        // the chain call, and possibly in an IMPLICIT_DEF coming from an
+        // llvm.amdgcn.dead intrinsic.
+        if (IsIWWFunction && (IsChainCall || IsImplicitDef) &&
+            TRI.isVectorRegister(MRI, Reg))
+          continue;
```
Doesn't matter if it's a vector register. I thought we already had a helper somewhere to skip synthetic uses

rovka committed:
Skip implicit defs unconditionally

jayfoad commented:
> Skip implicit defs unconditionally

I don't think it's safe to ignore all implicit def operands (but you can ignore the def operand of IMPLICIT_DEF). Do you need this to get the optimization you're hoping for? Is ignoring all use operands not enough?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUResourceUsageAnalysis.cpp:
```diff
@@ -239,6 +247,15 @@ AMDGPUResourceUsageAnalysis::analyzeResourceUsage(
           break;
         }
 
+        // For functions that use the llvm.amdgcn.init.whole.wave intrinsic, we
+        // often add artificial VGPR arguments for the purpose of preserving
+        // their inactive lanes. These should not be reported as part of our
+        // VGPR usage. We can identify them easily because they're only used in
+        // the chain call, and possibly in an IMPLICIT_DEF coming from an
+        // llvm.amdgcn.dead intrinsic.
+        if (IsIWWFunction && IsChainCall && TRI.isVectorRegister(MRI, Reg))
+          continue;
```
extend to any implicit argument that isn't in the MCInstrDesc, ignoring variadic instructions in the variadic section

rovka committed:
Skip register uses in AMDGPUResourceUsageAnalysis

Don't count register uses when determining the maximum number of
registers used by a function. Count only the defs. This is really an
underestimate of the true register usage, but in practice that's not
a problem because if a function uses a register, then it has either
defined it earlier, or some other function that executed before has
defined it. In particular, the register counts are used:
1. When launching an entry function - in which case we're safe because
   the register counts of the entry function will include the register
   counts of all callees.
2. At function boundaries in dynamic VGPR mode. In this case it's safe
   because whenever we set the new VGPR allocation we take into account
   the outgoing_vgpr_count set by the middle-end.

The main advantage of doing this is that the artificial VGPR arguments
used only for preserving the inactive lanes when using the
llvm.amdgcn.init.whole.wave intrinsic are no longer counted. This
enables us to allocate only the registers we need in dynamic VGPR mode.

rovka commented:
> > Skip implicit defs unconditionally
> 
> I don't think it's safe to ignore all implicit def operands (but you can ignore the def operand of IMPLICIT_DEF). Do you need this to get the optimization you're hoping for? Is ignoring all use operands not enough?

The code was only ignoring IMPLICIT_DEF instructions, not implicit def operands :) Sorry about the confusing commit message.

rovka commented:
> This could be much more generic. AMDGPUResourceUsageAnalysis could unconditionally ignore IMPLICIT_DEF and all use operands.

Ok, I switched to that. 

@arsenm I think this makes your other comments obsolete. But anyway, let me know what you think of this approach :)

rovka committed:
Merge branch 'main' into vgpr-count

Flakebi commented on llvm/docs/AMDGPUUsage.rst:
```diff
@@ -4255,10 +4255,9 @@ same *vendor-name*.
                                                                   GFX6-GFX9. A register
                                                                   is required if it is
-                                                                  used explicitly, or
+                                                                  written to, or
```
nit: reading from a register should count as well (e.g. a function that reads arguments from high VGPRs, or a shader that reads hw-initialized registers and only writes to lower VGPRs).

rovka replied:
I think @jayfoad's point was that we don't need to count those because they're counted wherever they're defined. I tried to explain that in the new PR description, but let me know if you disagree with that reasoning (personally I'm also leaning towards being more precise, i.e. including the reads as well, but then I don't know how to reduce the usage for the init-whole-wave case without giving it special treatment).

Flakebi commented on llvm/lib/Target/AMDGPU/AMDGPUAsmPrinter.cpp:
```diff
@@ -990,7 +990,10 @@ void AMDGPUAsmPrinter::getSIProgramInfo(SIProgramInfo &ProgInfo,
   unsigned WaveDispatchNumSGPR = 0, WaveDispatchNumVGPR = 0;
 
-  if (isShader(F.getCallingConv())) {
+  // Shaders that use the init.whole.wave intrinsic sometimes have VGPR
+  // arguments that are only added for the purpose of preserving their inactive
+  // lanes. Skip including them in the VGPR count.
+  if (isShader(F.getCallingConv()) && !MFI->hasInitWholeWave()) {
```
Why do we need to make this conditional on `hasInitWholeWave`?

I think ignoring implicit VGPRs should either apply to all shaders (i.e. hw entrypoints) and functions or – if it causes problems for shaders – apply to all (gpu-callable) functions, but not shaders/hw entrypoints.

rovka replied:
I think it might cause problems for entry functions when using wave dispatch (I got this from an commit message, I don't really know how that works). But good point that I can probably make it apply to all non-entry functions.

rovka committed:
s/no-init-whole-wave/isEntryFunction

tsymalla commented on llvm/test/CodeGen/AMDGPU/init-whole-wave-vgpr-count-use-inactive.ll:
```diff
@@ -0,0 +1,78 @@
+; CHECK-LABEL: .shader_functions:
+
+; Make sure that .vgpr_count doesn't include the %inactive.vgpr registers.
+; The shader is free to use any of the VGPRs mapped to a %inactive.vpgr as long as it only touches its active lanes.
```
```suggestion
; The shader is free to use any of the VGPRs mapped to a %inactive.vgpr as long as it only touches its active lanes.
```

rovka committed:
Fix typo in comment. NFC

Co-authored-by: Thomas Symalla <5754458+tsymalla@users.noreply.github.com>

Flakebi commented:
I created some small tests, to make sure this works as intended in all cases. Probably makes sense to add them here.
One of them yields surprising results, probably something that should be fixed.

PAL tests:
```llvm
; RUN: llc -mcpu=gfx1200 -o - < %s | FileCheck %s
; Check that reads of a VGPR in kernels counts towards VGPR count, but in functions, only writes of VGPRs count towards VGPR count.
target triple = "amdgcn--amdpal"

@global = addrspace(1) global i32 poison, align 4

; CHECK-LABEL: amdpal.pipelines:

; Neither uses not writes a VGPR, but the hardware initializes the VGPRs that the kernel receives, so they count as used.
; CHECK-LABEL: .entry_point_symbol: kernel_use
; CHECK: .vgpr_count:     0x20
define amdgpu_cs void @kernel_use([32 x i32] %args) {
entry:
  %a = extractvalue [32 x i32] %args, 14
  store i32 %a, ptr addrspace(1) @global
  ret void
}

; Neither uses not writes a VGPR
; CHECK-LABEL: gfx_func:
; CHECK: .vgpr_count:     0x20
define amdgpu_gfx [32 x i32] @gfx_func([32 x i32] %args) {
entry:
  ret [32 x i32] %args
}

; Neither uses not writes a VGPR
; CHECK-LABEL: chain_func:
; CHECK: .vgpr_count:     0x1
define amdgpu_cs_chain void @chain_func([32 x i32] %args) {
entry:
  call void (ptr, i32, {}, [32 x i32], i32, ...) @llvm.amdgcn.cs.chain.p0.i32.s.a(
        ptr @chain_func, i32 0, {} inreg {}, [32 x i32] %args, i32 0)
  unreachable
}
```

The (to me) surprising one is `gfx_func`, it only contains SALU instructions, so should have no defs of VGPRs and only uses for the return. I would expect it to have `vgpr_count: 0x0` or maybe 0x1.

This one works as expected:
```llvm
; RUN: llc -mcpu=gfx1200 -o - < %s | FileCheck %s
target triple = "amdgcn--amdpal"

declare amdgpu_gfx void @gfx_dummy([32 x i32] %args)

; CHECK-LABEL: .entry_point_symbol: kernel_call
; CHECK: .vgpr_count:     0x20
define amdgpu_cs void @kernel_call([32 x i32] %args) {
entry:
  call amdgpu_gfx void @gfx_dummy([32 x i32] %args)
  ret void
}
```

Carefully crafted compute test (the hw initializes at most one VGPR, so the test needs to ensure, no VGPR is ever written from any instruction). Also works as expected (correctly marks one VGPR as used).
```llvm
; RUN: llc -mcpu=gfx1200 -o - < %s | FileCheck %s
target triple = "amdgcn-amd-amdhsa"

@global = addrspace(1) global i32 poison, align 4

; Carefully crafted kernel that uses v0 but never writes a VGPR or reads another VGPR.
; Only hardware-initialized VGPRs (v0) are read in this kernel.

; CHECK-LABEL: amdhsa.kernels:
; CHECK: .vgpr_count:     1
define amdgpu_kernel void @kernel(ptr addrspace(8) %rsrc) #0 {
entry:
  %id = call i32 @llvm.amdgcn.workitem.id.x()
  call void @llvm.amdgcn.raw.ptr.buffer.store.i32(i32 %id, ptr addrspace(8) %rsrc, i32 0, i32 0, i32 0)
  ret void
}

attributes #0 = { "amdgpu-no-workitem-id-y" "amdgpu-no-workitem-id-z" }
```

rovka committed:
Unify code paths; include user + sys SGPRS if used

rovka committed:
Merge remote-tracking branch 'remotes/origin/main' into vgpr-count

rovka commented on llvm/test/CodeGen/AMDGPU/coalescer_remat.ll:
```diff
@@ -12,7 +12,7 @@ declare float @llvm.fma.f32(float, float, float)
 ; CHECK:  v_mov_b32_e32 v{{[0-9]+}}, 0
 ; It's probably OK if this is slightly higher:
-; CHECK: ; NumVgprs: 8
+; CHECK: ; NumVgprs: 5
```
I could use another pair of eyes on this test. From what I can tell, it uses v[4:7] (`buffer_store_dwordx4 v[4:7], off, s[0:3], 0`) but only defines v4. Am I missing smth, or is this just broken?

arsenm replied:
The first store in the loop is doing a vector store with only the 0th element defined

rovka reviewed (commented):
I made some pretty large changes:
* Unified handling of leaf and non-leaf functions as much as possible.
* Added any used preloaded SGPRs to the SGPR count (since they'll be written by the hardware)

Flakebi commented:
Thanks for the refactoring!

The tests I quoted above test specific corner cases, I find it much easier to look at and verify a 2 line test than a 20 line test.
The last test of the three I wrote fails with the new version of this change. It emits `.vgpr_count: 0` even though `v0` is live and defined as it’s initialized by the hardware (`v2` is never written in the shader).
I’d still like it if we can have the tests in llvm :)

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUResourceUsageAnalysis.cpp:
```diff
@@ -137,274 +137,29 @@ AMDGPUResourceUsageAnalysis::analyzeResourceUsage(
+    Info.NumAGPR = TRI.getNumDefinedPhysRegs(MRI, AMDGPU::AGPR_32RegClass);
+
+  // Count any user or system SGPRs that are actually used.
+  for (int I = MFI->getNumPreloadedSGPRs() - 1; I >= 0; I--)
```
Braces, but also should just take the raw number of preloaded SGPRs to start with. What happens if you don't allocate preloaded registers that were requested? If we wanted to trim out unused preloaded registers, it should have happened earlier (or, we wouldn't have preloaded in the first place). This could also break the debugger use case if it expects to find something there

arsenm commented on llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:
```diff
@@ -4053,6 +4053,21 @@ SIRegisterInfo::getNumUsedPhysRegs(const MachineRegisterInfo &MRI,
+
+  for (MCPhysReg Reg : reverse(RC.getRegisters()))
+    for (MCRegAliasIterator AI(Reg, this, true); AI.isValid(); ++AI)
+      if (!(MRI.def_empty(*AI) || std::all_of(MRI.def_begin(*AI), MRI.def_end(),
```
def_empty is redundant. Also use all_of(MRI.def_instructions()), you don't care about the specific def operands

arsenm commented on llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp:
```diff
@@ -4053,6 +4053,21 @@ SIRegisterInfo::getNumUsedPhysRegs(const MachineRegisterInfo &MRI,
+    return Op.getParent()->isImplicitDef();
+  };
+
+  for (MCPhysReg Reg : reverse(RC.getRegisters()))
```
Braces

arsenm commented on llvm/test/CodeGen/AMDGPU/init-whole-wave-vgpr-count-large.ll:
```diff
@@ -0,0 +1,76 @@
+
+!amdgpu.pal.metadata.msgpack = !{!0}
+
+!0 = !{!"\82\B0amdpal.pipelines\91\8B\A4.api\A6Vulkan\B2.compute_registers\85\AB.tg_size_en\C3\AA.tgid_x_en\C3\AA.tgid_y_en\C3\AA.tgid_z_en\C3\AF.tidig_comp_cnt\00\B0.hardware_stages\81\A3.cs\8D\AF.checksum_value\00\AB.debug_mode\00\AB.float_mode\CC\C0\A9.image_op\C2\AC.mem_ordered\C3\AB.sgpr_limitj\B7.threadgroup_dimensions\93 \01\01\AD.trap_present\00\B2.user_data_reg_map\90\AB.user_sgprs\10\AB ...
```
Can you avoid all this metadata? Also end of file line error

rovka committed:
Add missing tests. Fix preloaded VGPR issue

rovka committed:
Formatting

rovka committed:
Cound all preloaded regs

rovka committed:
Merge remote-tracking branch 'origin/main' into vgpr-count

Flakebi commented:
> So I would ask again, do we actually _need_ to report that one VGPR is used in Sebastian's carefully crafted compute test?

Not a definitive answer, but I think it would be cleaner to report v0 as used.
1. It allows to display the correct count in tools
2. and in case the hw behavior to always allocate at least one VGPR block changes, it still works correctly.

rovka commented:
> So you would distinguish between:
> 
> * Shader reads v0 which was initialized by hardware. This counts as "used".
> * Shader reads an undefined value from v0 which was not initialized by hardware. This does not count as "used".
> 
> I guess we can do that. It just seems like extra effort in the compiler for (I claim) no tangible benefit.

So are we good with the patch as it is now? It currently reports v0 as used if the workitem-id-x isn't explicitly disabled.

nhaehnle reviewed (approved):
I agree with @Flakebi that for entry functions it's better to report what the HW initializes.

jayfoad commented on llvm/lib/Target/AMDGPU/SIMachineFunctionInfo.h:
```diff
@@ -974,6 +974,19 @@ class SIMachineFunctionInfo final : public AMDGPUMachineFunction,
     return NumUserSGPRs + NumSystemSGPRs;
   }
 
+  unsigned getNumPreloadedVGPRs() const {
```
Does this need to handle all the other things that can get preloaded in VGPRs for graphics shaders? Or can you comment (or assert) that it is only for compute shaders?

Does this correctly handle the way that workitemidx/y/z are all packed into v0 in GFX11+?

rovka replied:
> Does this need to handle all the other things that can get preloaded in VGPRs for graphics shaders? Or can you comment (or assert) that it is only for compute shaders?

Graphics has special handling in AMDGPUAsmPrinter (so I think the final .vgpr_count should be correct). I'm tot sure if it's trivial to move it here. I'll have a look.

> Does this correctly handle the way that workitemidx/y/z are all packed into v0 in GFX11+?

Yes, there's a test for that in vgpr-count-compute.ll :)

rovka committed:
De Morgan

rovka committed:
Comments

rovka committed:
Merge remote-tracking branch 'remotes/origin/main' into vgpr-count-rovka

rovka committed:
Merge branch 'main' into vgpr-count
